src/v/cloud_storage/partition_manifest.cc

```diff
@@ -513,15 +514,24 @@ size_t partition_manifest::replaced_segments_count() const {
 }
 
 size_t partition_manifest::move_aligned_offset_range(
-  model::offset begin_inclusive, model::offset end_inclusive) {
+  model::offset begin_inclusive,
+  model::offset end_inclusive,
+  const segment_meta& replacing_segment) {
     size_t total_replaced_size = 0;
-
+    auto replacing_path = generate_segment_path(replacing_segment);
     auto it = _segments.lower_bound(begin_inclusive);
     while (it != _segments.end()
            // The segment is considered replaced only if all its
            // offsets are covered by new segment's offset range
            && it->second.base_offset >= begin_inclusive
            && it->second.committed_offset <= end_inclusive) {
+        if (generate_segment_path(it->second) == replacing_path) {
+            // The replacing segment shouldn't be exactly the same as the
+            // one that we already have in the manifest. Attempt to re-add
+            // same segment twice leads to data loss.
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format, "segment is already added {}", replacing_segment));
```

Comment:
Wondering what will happen when this throws: I think we'll not advance the in-sync offset and the waiters won't be notified, and eventually the operation will be timed out. But, the next time an op is replicated from the archiver loop, it will proceed.

Reply:
yes

Reply:
that make sense, I'll update the code